File: src/main/java/com/uiptv/service/CategoryService.java
package com.uiptv.service;

import com.uiptv.api.LoggerCallback;
import com.uiptv.db.CategoryDb;
import com.uiptv.db.SeriesCategoryDb;
import com.uiptv.db.VodCategoryDb;
import com.uiptv.model.Account;
import com.uiptv.model.Category;
import com.uiptv.shared.PlaylistEntry;
import com.uiptv.ui.LogDisplayUI;
import com.uiptv.ui.RssParser;
import com.uiptv.ui.XtremeParser;
import com.uiptv.util.AccountType;
import com.uiptv.util.FetchAPI;
import com.uiptv.util.ServerUtils;
import org.json.JSONArray;
import org.json.JSONObject;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.*;

import static com.uiptv.model.Account.AccountAction.itv;
import static com.uiptv.model.Account.AccountAction.series;
import static com.uiptv.model.Account.AccountAction.vod;
import static com.uiptv.model.Account.NOT_LIVE_TV_CHANNELS;
import static com.uiptv.util.AccountType.*;
import static com.uiptv.util.FetchAPI.nullSafeBoolean;
import static com.uiptv.util.FetchAPI.nullSafeInteger;
import static com.uiptv.util.M3U8Parser.parsePathCategory;
import static com.uiptv.util.M3U8Parser.parseUrlCategory;
import static com.uiptv.widget.UIptvAlert.showError;

public class CategoryService {
    private static final long VOD_SERIES_CACHE_TTL_MS = 30L * 24L * 60L * 60L * 1000L;
    private static CategoryService instance;
    private final ContentFilterService contentFilterService;

    private CategoryService() {
        this.contentFilterService = ContentFilterService.getInstance();
    }

    public static synchronized CategoryService getInstance() {
        if (instance == null) {
            instance = new CategoryService();
        }
        return instance;
    }

    private static Map<String, String> getCategoryParams(Account.AccountAction accountAction) {
        final Map<String, String> params = new HashMap<>();
        params.put("JsHttpRequest", new Date().getTime() + "-xml");
        params.put("type", accountAction.name());
        params.put("action", accountAction == itv ? "get_genres" : "get_categories");
        return params;
    }

    public List<Category> get(Account account) {
        return get(account, true);
    }

    public List<Category> get(Account account, boolean censor) {
        return get(account, censor, null);
    }

    public List<Category> get(Account account, boolean censor, LoggerCallback logger) {
        if (account.getType() == RSS_FEED) {
            hardReloadCategories(account, logger);
            List<Category> cats = CategoryDb.get().getCategories(account);
            return maybeFilterCategories(cats, censor);
        }

        if ((account.getAction() == vod || account.getAction() == series)
                && (account.getType() == STALKER_PORTAL || account.getType() == XTREME_API)) {
            List<Category> cached = getVodSeriesCachedCategories(account);
            if (!cached.isEmpty() && isVodSeriesCategoriesFresh(account)) {
                log(logger, "Loaded categories from local cache.");
                return maybeFilterCategories(cached, censor);
            }

            log(logger, "No fresh cached categories found. Fetching from portal/provider...");
            List<Category> fetched = fetchCategoriesFromSource(account);
            if (!fetched.isEmpty()) {
                saveVodSeriesCategories(account, fetched);
                log(logger, "Saved " + fetched.size() + " categories to local VOD/Series cache.");
                List<Category> stored = getVodSeriesCachedCategories(account);
                return maybeFilterCategories(stored.isEmpty() ? fetched : stored, censor);
            }
            return maybeFilterCategories(cached, censor);
        }

        if (NOT_LIVE_TV_CHANNELS.contains(account.getAction())) {
            if (account.getType() == STALKER_PORTAL || account.getType() == XTREME_API) {
                List<Category> cachedCategories = CategoryDb.get().getCategories(account);
                if (!cachedCategories.isEmpty()) {
                    log(logger, "Loaded categories from local cache.");
                    return maybeFilterCategories(cachedCategories, censor);
                }

                log(logger, "No cached categories found. Fetching from portal/provider...");
                List<Category> fetchedCategories = fetchCategoriesFromBackend(account, logger);
                if (!fetchedCategories.isEmpty()) {
                    CategoryDb.get().saveAll(fetchedCategories, account);
                    log(logger, "Saved " + fetchedCategories.size() + " categories to local cache.");
                }
                return maybeFilterCategories(fetchedCategories, censor);
            }

            hardReloadCategories(account, logger);
            List<Category> cats = CategoryDb.get().getCategories(account);
            return maybeFilterCategories(cats, censor);
        }

        List<Category> cachedCategories = CategoryDb.get().getCategories(account);
        if (cachedCategories.isEmpty() || account.getAction() != itv) {
            hardReloadCategories(account, logger);
            cachedCategories.addAll(CategoryDb.get().getCategories(account));
        }
        return maybeFilterCategories(cachedCategories, censor);
    }

    private List<Category> fetchCategoriesFromSource(Account account) {
        List<Category> categories = new ArrayList<>();
        try {
            if (account.getType() == AccountType.XTREME_API) {
                categories.addAll(xtremeAPICategories(account));
            } else {
                List<Category> s = stalkerPortalCategories(account, null);
                if (s != null && !s.isEmpty()) categories.addAll(s);
            }
        } catch (Exception e) {
            LogDisplayUI.addLog("Network Error: " + e.getMessage());
        }
        return categories;
    }

    private List<Category> getVodSeriesCachedCategories(Account account) {
        if (account.getAction() == vod) {
            return VodCategoryDb.get().getCategories(account);
        }
        if (account.getAction() == series) {
            return SeriesCategoryDb.get().getCategories(account);
        }
        return Collections.emptyList();
    }

    private boolean isVodSeriesCategoriesFresh(Account account) {
        if (account.getAction() == vod) {
            return VodCategoryDb.get().isFresh(account, VOD_SERIES_CACHE_TTL_MS);
        }
        if (account.getAction() == series) {
            return SeriesCategoryDb.get().isFresh(account, VOD_SERIES_CACHE_TTL_MS);
        }
        return false;
    }

    private void saveVodSeriesCategories(Account account, List<Category> categories) {
        if (account.getAction() == vod) {
            VodCategoryDb.get().saveAll(categories, account);
        } else if (account.getAction() == series) {
            SeriesCategoryDb.get().saveAll(categories, account);
        }
    }

    private void hardReloadCategories(Account account) {
        hardReloadCategories(account, null);
    }

    private void hardReloadCategories(Account account, LoggerCallback logger) {
        List<Category> categories = fetchCategoriesFromBackend(account, logger);
        CategoryDb.get().saveAll(categories, account);
    }

    private List<Category> fetchCategoriesFromBackend(Account account) {
        return fetchCategoriesFromBackend(account, null);
    }

    private List<Category> fetchCategoriesFromBackend(Account account, LoggerCallback logger) {
        List<Category> categories = new ArrayList<>();
        try {
            if (Objects.requireNonNull(account.getType()) == AccountType.M3U8_LOCAL || account.getType() == M3U8_URL) {
                categories.addAll(m3u8Categories(account));
            } else if (account.getType() == AccountType.XTREME_API) {
                log(logger, "Fetching categories from Xtreme API...");
                categories.addAll(xtremeAPICategories(account));
            } else if (account.getType() == AccountType.RSS_FEED) {
                categories.addAll(rssCategories());
            } else {
                log(logger, "Fetching categories from Stalker Portal...");
                List<Category> s = stalkerPortalCategories(account, logger);
                if (s != null && !s.isEmpty()) categories.addAll(s);
            }
        } catch (Exception e) {
            LogDisplayUI.addLog("Network Error: " + e.getMessage());
            log(logger, "Network error while loading categories: " + e.getMessage());
        }
        return categories;
    }

    private List<Category> xtremeAPICategories(Account account) {
        return XtremeParser.parseCategories(account);
    }

    private List<Category> m3u8Categories(Account account) throws MalformedURLException {
        Set<Category> categories = new LinkedHashSet<>();
        Set<PlaylistEntry> m3uEntries = account.getType() == M3U8_URL ? parseUrlCategory(new URL(account.getM3u8Path())) : parsePathCategory(account.getM3u8Path());
        m3uEntries.forEach(entry -> {
            Category c = new Category(entry.getId(), entry.getGroupTitle(), entry.getGroupTitle(), false, 0);
            categories.add(c);
        });
        return categories.stream().toList();
    }

    private List<Category> rssCategories() {
        Set<Category> categories = new LinkedHashSet<>();
        Set<PlaylistEntry> m3uEntries = RssParser.getCategories();
        m3uEntries.forEach(entry -> {
            Category c = new Category(entry.getId(), entry.getGroupTitle(), entry.getGroupTitle(), false, 0);
            categories.add(c);
        });
        return categories.stream().toList();
    }

    private List<Category> stalkerPortalCategories(Account account, LoggerCallback logger) {
        log(logger, "Performing portal handshake...");
        HandshakeService.getInstance().connect(account);
        if (account.isNotConnected()) {
            log(logger, "Handshake failed.");
            return null;
        }
        log(logger, "Handshake successful. Loading categories...");
        String jsonResponse = FetchAPI.fetch(getCategoryParams(account.getAction()), account);
        return parseCategories(jsonResponse, false);
    }

    public String readToJson(Account account) {
        return ServerUtils.objectToJson(get(account));
    }

    public List<Category> parseCategories(String json, boolean censor) {
        List<Category> categoryList = new ArrayList<>();
        try {
            JSONArray list = new JSONObject(json).getJSONArray("js");
            for (int i = 0; i < list.length(); i++) {
                JSONObject jsonCategory = list.getJSONObject(i);
                Category category = new Category(jsonCategory.getString("id"), jsonCategory.getString("title"), jsonCategory.getString("alias"), nullSafeBoolean(jsonCategory, "active_sub"), nullSafeInteger(jsonCategory, "censored"));
                category.setExtraJson(jsonCategory.toString());
                categoryList.add(category);
            }
        } catch (Exception e) {
            showError("Error while processing response data" + e.getMessage());
        }
        return maybeFilterCategories(categoryList, censor);
    }

    private List<Category> maybeFilterCategories(List<Category> categories, boolean applyFilter) {
        return applyFilter ? contentFilterService.filterCategories(categories) : categories;
    }

    private void log(LoggerCallback logger, String message) {
        if (logger != null) {
            logger.log(message);
        }
    }

}

File: server/src/main/kotlin/com/uiptv/service/CategoryService.kt
package com.uiptv.service

import com.uiptv.api.LoggerCallback
import com.uiptv.db.CategoryDb
import com.uiptv.db.SeriesCategoryDb
import com.uiptv.db.VodCategoryDb
import com.uiptv.model.Account
import com.uiptv.model.Category
import com.uiptv.model.CategoryType
import com.uiptv.shared.PlaylistEntry
import com.uiptv.util.AccountType
import com.uiptv.util.AppLog
import com.uiptv.util.FetchAPI
import com.uiptv.util.RssParser
import com.uiptv.util.ServerUtils
import com.uiptv.util.XtremeApiParser
import com.uiptv.util.json.asJsonString
import com.uiptv.util.json.optArray
import com.uiptv.util.json.optObject
import com.uiptv.util.json.parseJsonObject
import java.net.MalformedURLException
import java.net.URL
import java.util.Date
import java.util.LinkedHashSet
import com.uiptv.model.Account.AccountAction.itv
import com.uiptv.model.Account.AccountAction.series
import com.uiptv.model.Account.AccountAction.vod

class CategoryService(
    private val contentFilterService: ContentFilterService,
    private val configurationService: ConfigurationService,
    private val handshakeService: HandshakeService
) {
    constructor() : this(ContentFilterService, ConfigurationService, HandshakeService.INSTANCE)


    fun get(account: Account): List<Category> = get(account, true)
    fun getCached(account: Account?): List<Category> {
        if (account == null) {
            return emptyList()
        }
        return when (account.action) {
            vod -> VodCategoryDb.get().getCategories(account)
            series -> SeriesCategoryDb.get().getCategories(account)
            else -> CategoryDb.get().getCategories(account)
        }
    }
    fun get(account: Account, censor: Boolean): List<Category> = get(account, censor, null)
    fun get(account: Account, censor: Boolean, logger: LoggerCallback?): List<Category> {
        if (account.type == AccountType.RSS_FEED) {
            hardReloadCategories(account, logger)
            return maybeFilterCategories(CategoryDb.get().getCategories(account), censor)
        }

        if (usesVodSeriesCategoryCache(account)) {
            return getVodSeriesCategories(account, censor, logger)
        }

        if (Account.NOT_LIVE_TV_CHANNELS.contains(account.action)) {
            if (account.type == AccountType.STALKER_PORTAL || account.type == AccountType.XTREME_API) {
                val cachedCategories = CategoryDb.get().getCategories(account)
                if (cachedCategories.isNotEmpty()) {
                    log(logger, "Loaded categories from local cache.")
                    return maybeFilterCategories(cachedCategories, censor)
                }
                log(logger, "No cached categories found. Fetching from portal/provider...")
                val fetchedCategories = fetchCategoriesFromBackend(account, logger)
                if (fetchedCategories.isNotEmpty()) {
                    CategoryDb.get().saveAll(fetchedCategories, account)
                    log(logger, "Saved ${fetchedCategories.size} categories to local cache.")
                }
                return maybeFilterCategories(fetchedCategories, censor)
            }

            hardReloadCategories(account, logger)
            return maybeFilterCategories(CategoryDb.get().getCategories(account), censor)
        }

        val cachedCategories = CategoryDb.get().getCategories(account).toMutableList()
        if (cachedCategories.isEmpty() || account.action != itv) {
            hardReloadCategories(account, logger)
            cachedCategories.addAll(CategoryDb.get().getCategories(account))
        }
        return maybeFilterCategories(cachedCategories, censor)
    }

    private fun usesVodSeriesCategoryCache(account: Account): Boolean =
        (account.action == vod || account.action == series) &&
            (account.type == AccountType.STALKER_PORTAL || account.type == AccountType.XTREME_API)

    private fun getVodSeriesCategories(account: Account, censor: Boolean, logger: LoggerCallback?): List<Category> {
        val cached = getVodSeriesCachedCategories(account)
        if (cached.isNotEmpty() && isVodSeriesCategoriesFresh(account)) {
            log(logger, "Loaded categories from local cache.")
            return maybeFilterCategories(cached, censor)
        }
        log(logger, "No fresh cached categories found. Fetching from portal/provider...")
        val fetched = fetchCategoriesFromSource(account)
        if (fetched.isNotEmpty()) {
            saveVodSeriesCategories(account, fetched)
            log(logger, "Saved ${fetched.size} categories to local VOD/Series cache.")
            val stored = getVodSeriesCachedCategories(account)
            return maybeFilterCategories(if (stored.isEmpty()) fetched else stored, censor)
        }
        return maybeFilterCategories(cached, censor)
    }

    private fun fetchCategoriesFromSource(account: Account): List<Category> {
        val categories = ArrayList<Category>()
        try {
            if (account.type == AccountType.XTREME_API) {
                categories.addAll(xtremeAPICategories(account))
            } else {
                val stalker = stalkerPortalCategories(account, null)
                if (!stalker.isNullOrEmpty()) {
                    categories.addAll(stalker)
                }
            }
        } catch (e: Exception) {
            AppLog.addWarningLog(CategoryService::class.java, "Network Error: ${e.message}")
        }
        return categories
    }

    private fun getVodSeriesCachedCategories(account: Account): List<Category> =
        when (account.action) {
            vod -> VodCategoryDb.get().getCategories(account)
            series -> SeriesCategoryDb.get().getCategories(account)
            else -> emptyList()
        }

    private fun isVodSeriesCategoriesFresh(account: Account): Boolean {
        val cacheTtlMs = configurationService.getCacheExpiryMs()
        return when (account.action) {
            vod -> VodCategoryDb.get().isFresh(account, cacheTtlMs)
            series -> SeriesCategoryDb.get().isFresh(account, cacheTtlMs)
            else -> false
        }
    }

    private fun saveVodSeriesCategories(account: Account, categories: List<Category>) {
        when (account.action) {
            vod -> VodCategoryDb.get().saveAll(categories, account)
            series -> SeriesCategoryDb.get().saveAll(categories, account)
            else -> {}
        }
    }

    private fun hardReloadCategories(account: Account, logger: LoggerCallback?) {
        CategoryDb.get().saveAll(fetchCategoriesFromBackend(account, logger), account)
    }

    private fun fetchCategoriesFromBackend(account: Account, logger: LoggerCallback?): List<Category> {
        val categories = ArrayList<Category>()
        try {
            when (account.type) {
                AccountType.M3U8_LOCAL, AccountType.M3U8_URL -> categories.addAll(m3u8Categories(account))
                AccountType.XTREME_API -> {
                    log(logger, "Fetching categories from Xtreme API...")
                    categories.addAll(xtremeAPICategories(account))
                }
                AccountType.RSS_FEED -> categories.addAll(rssCategories())
                else -> {
                    log(logger, "Fetching categories from Stalker Portal...")
                    val stalker = stalkerPortalCategories(account, logger)
                    if (!stalker.isNullOrEmpty()) {
                        categories.addAll(stalker)
                    }
                }
            }
        } catch (e: Exception) {
            AppLog.addWarningLog(CategoryService::class.java, "Network Error: ${e.message}")
            log(logger, "Network error while loading categories: ${e.message}")
        }
        return categories
    }

    private fun xtremeAPICategories(account: Account): List<Category> = XtremeApiParser.parseCategories(account)

    @Throws(MalformedURLException::class)
    private fun m3u8Categories(account: Account): List<Category> {
        val categories = LinkedHashSet<Category>()
            val m3uEntries: Set<PlaylistEntry> =
            if (account.type == AccountType.M3U8_URL) {
                com.uiptv.util.M3U8Parser.parseUrlCategory(com.uiptv.util.UiptUtils.parseUrlLikeUri(account.m3u8Path.orEmpty()).toURL())
            } else {
                com.uiptv.util.M3U8Parser.parsePathCategory(account.m3u8Path.orEmpty())
            }
        m3uEntries.forEach { entry ->
            categories.add(Category(entry.id, entry.groupTitle, entry.groupTitle, false, 0))
        }
        return categories.toList()
    }

    private fun rssCategories(): List<Category> {
        val categories = LinkedHashSet<Category>()
        RssParser.getCategories().forEach { entry ->
            categories.add(Category(entry.id, entry.groupTitle, entry.groupTitle, false, 0))
        }
        return categories.toList()
    }

    private fun stalkerPortalCategories(account: Account, logger: LoggerCallback?): List<Category> {
        log(logger, "Performing portal handshake...")
        handshakeService.connect(account)
        if (account.isNotConnected()) {
            log(logger, "Handshake failed.")
            return emptyList()
        }
        log(logger, "Handshake successful. Loading categories...")
        return parseCategories(FetchAPI.fetch(getCategoryParams(account.action), account), false)
    }
    fun readToJson(account: Account): String = ServerUtils.objectToJson(get(account))
    fun parseCategories(json: String?, censor: Boolean): List<Category> {
        val categoryList = ArrayList<Category>()
        try {
            val root = parseJsonObject(json) ?: return maybeFilterCategories(categoryList, censor)
            val list = root.optArray("js")
                ?: root.optObject("js")?.optArray("data")
                ?: root.optArray("data")
                ?: return maybeFilterCategories(categoryList, censor)
            for (index in list.indices) {
                val jsonCategory = list.optObject(index) ?: continue
                val categoryId = firstNonBlank(
                    FetchAPI.nullSafeString(jsonCategory, "id"),
                    FetchAPI.nullSafeString(jsonCategory, "category_id"),
                    FetchAPI.nullSafeString(jsonCategory, "tv_genre_id"),
                    FetchAPI.nullSafeString(jsonCategory, "alias"),
                    FetchAPI.nullSafeString(jsonCategory, "title"),
                    FetchAPI.nullSafeString(jsonCategory, "name")
                )
                val title = firstNonBlank(
                    FetchAPI.nullSafeString(jsonCategory, "title"),
                    FetchAPI.nullSafeString(jsonCategory, "name"),
                    categoryId
                )
                val alias = firstNonBlank(FetchAPI.nullSafeString(jsonCategory, "alias"), title, categoryId)
                if (categoryId.isBlank() || title.isBlank()) continue
                val category = Category(
                    categoryId,
                    title,
                    alias,
                    FetchAPI.nullSafeBoolean(jsonCategory, "active_sub"),
                    FetchAPI.nullSafeInteger(jsonCategory, "censored")
                )
                category.extraJson = jsonCategory.asJsonString()
                categoryList.add(category)
            }
        } catch (e: Exception) {
            AppLog.addErrorLog(CategoryService::class.java, "Error while processing response data${e.message}")
        }
        return maybeFilterCategories(categoryList, censor)
    }

    private fun maybeFilterCategories(categories: List<Category>, applyFilter: Boolean): List<Category> {
        val normalized = normalizeAllAndUncategorized(categories)
        return if (applyFilter) contentFilterService.filterCategories(normalized) ?: normalized else normalized
    }

    private fun normalizeAllAndUncategorized(categories: List<Category>?): List<Category> {
        if (categories.isNullOrEmpty() || categories.size != 2) {
            return categories ?: emptyList()
        }
        val hasAll = categories.any { titleEquals(it, CategoryType.ALL.displayName()) }
        val hasUncategorized = categories.any { titleEquals(it, CategoryType.UNCATEGORIZED.displayName()) }
        if (!hasAll || !hasUncategorized) {
            return categories
        }
        return categories.filterNot { titleEquals(it, CategoryType.UNCATEGORIZED.displayName()) }
    }

    private fun titleEquals(category: Category?, value: String?): Boolean {
        val title = category?.title
        if (title == null || value == null) {
            return false
        }
        return title.trim().equals(value.trim(), true)
    }

    private fun firstNonBlank(vararg values: String?): String =
        values.firstOrNull { !it.isNullOrBlank() }?.trim().orEmpty()

    private fun log(logger: LoggerCallback?, message: String) {
        logger?.log(message)
    }

    private fun getCategoryParams(accountAction: Account.AccountAction): Map<String, String> {
        val params = HashMap<String, String>()
        params["JsHttpRequest"] = "${Date().time}-xml"
        params["type"] = accountAction.name
        params["action"] = if (accountAction == itv) "get_genres" else "get_categories"
        return params
    }

    companion object {
        @JvmField
        val INSTANCE: CategoryService = CategoryService(ContentFilterService, ConfigurationService, HandshakeService.INSTANCE)
    }
}
